### PR TITLE
Add demonic gorilla plugin

### PR DESCRIPTION
**Pull request by WooxSolo**
Added a demonic gorillas plugin which counts their attacks and shows what attack styles are possible for them to use next. It also looks for NPC movement patterns to detect when a gorilla is trying to melee and when it's trying to attack from a distance to show the appropriate attack styles that it might use.

This pull also includes hooks for hitsplats, hooks for overhead icons and a WorldArea class which is used to calculate line of sight and NPC movement prediction as those are used by the plugin.

image

image

### CODE REVIEW

File: runescape-client/src/main/java/Actor.java [hunk 1]
```diff
@@ -374,6 +375,8 @@ final void method1655() {
       signature = "(IIIIIII)V",
       garbageValue = "1600365055"
    )
+   @Export("applyActorHitsplat")
+   @Hook(value = "onActorHitsplat", end = true)
    final void method1657(int var1, int var2, int var3, int var4, int var5, int var6) {
```
Comment (Adam-) [inline runescape-client/src/main/java/Actor.java]: This method should be renamed to applyActorHitsplat

File: runelite-client/src/main/java/net/runelite/client/plugins/demonicgorilla/DemonicGorillaPlugin.java [hunk 2]
```diff
@@ -0,0 +1,663 @@
...
+		}
+	}
+
+	@Subscribe
+	public void onProjectile(ProjectileMoved event)
+	{
+		Projectile projectile = event.getProjectile();
+
+		// We only want to process projectiles that we already haven't processed
+		if (!previousProjectiles.contains(projectile))
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/demonicgorilla/DemonicGorillaPlugin.java]: You can return before this if the id is not a boulder/attack.

File: runelite-client/src/main/java/net/runelite/client/callback/Hooks.java [hunk 3]
```diff
@@ -434,4 +435,29 @@ public static void postItemComposition(ItemComposition itemComposition)
...
+	 *
+	 * @param actor The actor the hitsplat was applied to
+	 * @param value The value of the hitsplat (i.e. how high the hit was)
+	 * @param type The hitsplat type (i.e. color)
+	 * @param var3
+	 * @param var4
+	 * @param gameCycle The gamecycle the hitsplat was applied on
+	 * @param duration The amount of gamecycles the hitsplat will last for
+	 */
+	public static void onActorHitsplat(Actor actor, int type, int value, int var3, int var4,
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/callback/Hooks.java]: argument order here doesn't match javadoc

File: runelite-client/src/main/java/net/runelite/client/callback/Hooks.java [hunk 4]
```diff
@@ -102,7 +103,7 @@ public static void clientMainLoop(Client client, boolean arg1)
 		if (shouldProcessGameTick)
 		{
 			shouldProcessGameTick = false;
-			eventBus.post(tick);
+			eventBus.post(new GameTick());
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/callback/Hooks.java]: I think the previous logic of keeping one tick and just posting that is fine

File: runelite-api/src/main/java/net/runelite/api/CollisionDataFlag.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+	public static final int BREAK_LINE_OF_SIGHT_SOUTH = 0x4000;
+
+	// 0x100 for objects, 0x200000 for unwalkable tiles such as water etc,
+	// 0x40000 is very rare but Barbarian Assault cannon has it.
+	public static final int BREAK_MOVEMENT_FULL = 0x100 | 0x200000 | 0x40000;
+
+	public static final int BREAK_MOVEMENT_EAST = 0x8;
+	public static final int BREAK_MOVEMENT_WEST = 0x80;
+	public static final int BREAK_MOVEMENT_NORTH = 0x2;
+	public static final int BREAK_MOVEMENT_SOUTH = 0x20;
```
Comment (Adam-) [inline runelite-api/src/main/java/net/runelite/api/CollisionDataFlag.java]: I have some existing unreleased work on these:
```
int BLOCK_NORTH_WEST = 0x1; 
int BLOCK_NORTH = 0x2; 
int BLOCK_NORTH_EAST = 0x4; 
int BLOCK_EAST = 0x8; 
int BLOCK_SOUTH_EAST = 0x10;
int BLOCK_SOUTH = 0x20;
int BLOCK_SOUTH_WEST = 0x40;
int BLOCK_WEST = 0x80;
int OBJ = 0x100; 
int FLOORDECO = 0x40000;
int FLOOR = 0x200000;
```
Additionally if you shift any of the BLOCK_ masks by << 9  you get the mask for whether it blocks projectiles.

**Comment by Adam-** [comment]
Additionally, even though the current practice for plugins like this is to do npc/player scans per-tick, I think it would be worthwhile to make events for players/npcs spawning in and out and making the plugins use that instead.

File: runelite-api/src/main/java/net/runelite/api/CollisionDataFlag.java (hunk 5 above)
Comment (WooxSolo) [reply runelite-api/src/main/java/net/runelite/api/CollisionDataFlag.java]:
```
int BLOCK_NORTH_WEST = 0x1; 
int BLOCK_NORTH_EAST = 0x4; 
int BLOCK_SOUTH_EAST = 0x10;
int BLOCK_SOUTH_WEST = 0x40;
```
What are these used for? I got the line of sight algorithm from Henke, so he has probably tested it more than I have, but from what I've seen the algorithm works fine without those flags.

Comment (Bartvhelvert) [reply runelite-api/src/main/java/net/runelite/api/CollisionDataFlag.java]: They are used by the server sided pathing algorithm. In earlier versions of RS2 pathing was done in the client, that's where they come from.

File: runelite-client/src/main/java/net/runelite/client/callback/Hooks.java (hunk 3 above)
Comment (WooxSolo) [reply runelite-client/src/main/java/net/runelite/client/callback/Hooks.java]: Fixed this by moving the params in the doc

**Comment by WooxSolo** [comment]
I added line of sight and valid actor movement to the devtools as well

**Comment by henke96** [comment]
This looks wrong: public static final int BLOCK_LINE_OF_SIGHT_WEST = BLOCK_MOVEMENT_WEST << 0; // 0x10000
And no credit for los algorithm? :p

File: runelite-api/src/main/java/net/runelite/api/coords/WorldArea.java [hunk 6]
```diff
@@ -0,0 +1,631 @@
...
+import lombok.Getter;
+import net.runelite.api.Client;
+import net.runelite.api.CollisionData;
+import net.runelite.api.CollisionDataFlag;
+import net.runelite.api.Point;
+import net.runelite.api.Tile;
+
+public class WorldArea
+{
+	private static final int REGION_SIZE = 104;
```
Comment (Adam-) [inline runelite-api/src/main/java/net/runelite/api/coords/WorldArea.java]: There is a Constants.REGION_SIZE

File: runelite-client/src/main/java/net/runelite/client/plugins/demonicgorilla/DemonicGorillaPlugin.java [hunk 7]
```diff
@@ -0,0 +1,706 @@
...
+		{
+			reset();
+		}
+	}
+
+	@Subscribe
+	public void onPlayerSpawned(PlayerSpawned event)
+	{
+		Player player = event.getPlayer();
+		memorizedPlayers.put(player, new MemorizedPlayer(player));
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/demonicgorilla/DemonicGorillaPlugin.java]: I do not like the overhead of wrapping + tracking hitsplats etc per player all the time. Can it be limited to only if there are gorillas around? Or maybe by region?

File: runelite-client/src/main/java/net/runelite/client/plugins/demonicgorilla/DemonicGorillaPlugin.java [hunk 8]
```diff
@@ -0,0 +1,706 @@
...
+	}
+
+	private void clearFinishedProjectiles()
+	{
+		List<Projectile> allProjectiles = client.getProjectiles();
+		Iterator<Projectile> it = previousProjectiles.iterator();
+		while (it.hasNext())
+		{
+			Projectile projectile = it.next();
+			if (!allProjectiles.contains(projectile))
```
Comment (Adam-) [inline runelite-client/src/main/java/net/runelite/client/plugins/demonicgorilla/DemonicGorillaPlugin.java]: Projectiles have an `getRemainingCycles()` you can use instead.

**Pull request merged**